### PR TITLE
fix: Ensure that the constraints are checked and propagated fully

### PR DESCRIPTION
**Pull request by Marwes**
### fix: Always query the root variable for the constraint it is bound by

Only the root variable knows which kinds it is constrained by as we remove them and combine
them when two variables are unified

### fix: Always add constraints to the root variable

Else we aren't able to reliably lookup all constraints for a variable as they
may be spread out over multiple (otherwise unified) variables.

Extracted from #4776

### CODE REVIEW

**Review by nathanielc — CHANGES_REQUESTED** [review]
Can you add a test case for this case?

> Else we aren't able to reliably lookup all constraints for a variable as they
may be spread out over multiple (otherwise unified) variables.

I have some questions below to help me understand.

**Comment by Marwes** [comment]
> Can you add a test case for this case?

Fixed. Messed up the cherry-pick/rebase and dropped the test and a change.

File: libflux/flux-core/src/semantic/infer.rs [hunk 1]
```diff
@@ -182,6 +182,28 @@ pub fn equal(
     })
 }
 
+pub fn subsume(
```
[the remark this answers is not in the document]
Comment (Marwes) [reply libflux/flux-core/src/semantic/infer.rs:185]: > What is the intent of the name subsume? Is this related to subtyping somehow?

Yes, this change is for `fix: Don't cause errors in default argument checking due to label requirements`. I needed to change the check to use subsumption, else the default value (a string literal) would not get treated as a label and cause an error as the function required a label to be passed. It should probably have been a separate PR as it is a different change from fully checking the constraints fully.

>  Is root type the supertype in this context?

No. If multiple variables are unified the root is the variable that names the entire set.
Say you unify like this
```
A <> B
B <> C

D <> E
```
The set `A, B, C` may have the root `A` while `D, E` has the root `E`. Querying the root for `B` or `C` returns `A` in both cases which lets us know that they are unified while querying it for `D` will yield `E` so we know it is in a disjoint set from `B`, `C` (and `A`). Which variable in a set that ends up as the root is arbitrary and just a by product of how https://en.wikipedia.org/wiki/Disjoint-set_data_structure is implemented.

Specifically in this change I made sure that add a constraint on, for example, `B` (which has `A` as the root) we actually record the constraint on `A`. Before, when we added the constraint on `B` explicitly inference did not know that `A` and `C` were also constrained by this (hence lost).

[the remark this answers is not in the document]
Comment (nathanielc) [reply libflux/flux-core/src/semantic/infer.rs:185]: Thanks for the explanation, helps clarify.

**Review by nathanielc — APPROVED** [review]

**Pull request merged**